## Hug *et. al*

Ribosomal proteins L2, L3, L4, L5, L6, L14, L15, L16, L18, L22, L24, S3, S8, S10, S17, and S19.

- Hug, LA, Baker, BJ, Anantharaman, K, Brown, CT, Probst, AJ, Castelle, CJ, Butterfield, CN, Hernsdorf, AW, Amano, Y, Ise, K, Suzuki, Y, Dudek, N, Relman, DA, Finstad, KM, Amundson, R, Thomas, BC, Banfield, JF **(2016)**. A new view of the tree of life. *Nat Microbiol*, 1:16048. doi:[10.1038/nmicrobiol.2016.48](https://doi.org/10.1038/nmicrobiol.2016.48)
- Hug, LA, Castelle, CJ, Wrighton, KC, Thomas, BC, Sharon, I, Frischkorn, KR, Williams, KH, Tringe, SG, Banfield, JF **(2013)**. Community genomic analyses constrain the distribution of metabolic traits across the Chloroflexi phylum and indicate roles in sediment carbon cycling. *Microbiome*, 1, 1:22. doi:[10.1186/2049-2618-1-22](https://dx.doi.org/10.1186/2049-2618-1-22)

Pfam search results, cross checked by looking at the *E. coli* ribosomal proteins in UniProtKB (which links back to Pfam):

| ID | UniProtKB | Pfam |
| -- | -- | -- |
| L2 | https://www.uniprot.org/uniprot/P60422 | https://pfam.xfam.org/family/PF00181 |
| L3 | https://www.uniprot.org/uniprot/P60438 | https://pfam.xfam.org/family/PF00297 |
| L4 | https://www.uniprot.org/uniprot/P60723 |  https://pfam.xfam.org/family/PF00573 |
| L5 | https://www.uniprot.org/uniprot/P62399 | https://pfam.xfam.org/family/PF00281 |
| L6 | https://www.uniprot.org/uniprot/P0AG55 | https://pfam.xfam.org/family/PF00347 |
| L14 | https://www.uniprot.org/uniprot/P0ADY3 | https://pfam.xfam.org/family/PF00238 |
| L15 | https://www.uniprot.org/uniprot/P02413 | https://pfam.xfam.org/family/PF00828 |
| L16 | https://www.uniprot.org/uniprot/P0ADY7 | https://pfam.xfam.org/family/PF00252 |
| L18 | https://www.uniprot.org/uniprot/P0C018 | https://pfam.xfam.org/family/PF00861 |
| L22 | https://www.uniprot.org/uniprot/P61175 | https://pfam.xfam.org/family/PF00237 |
| L24 | https://www.uniprot.org/uniprot/P60624 | https://pfam.xfam.org/family/PF17136 |
| S3 | https://www.uniprot.org/uniprot/P0A7V3 | https://pfam.xfam.org/family/PF00189 |
| S8 | https://www.uniprot.org/uniprot/P0A7W7 | https://pfam.xfam.org/family/PF00410 |
| S10 | https://www.uniprot.org/uniprot/P0A7R5 | https://pfam.xfam.org/family/PF00338 |
| S17 | https://www.uniprot.org/uniprot/P0AG63 | https://pfam.xfam.org/family/PF00366 |
| S19 | https://www.uniprot.org/uniprot/P0A7U3 | https://pfam.xfam.org/family/PF00203 |


## PhyloSift

Set of 37 "elite" gene families previously identified as nearly universal and present in single-copy (subset of the 40 gene families described in Wu *et al.*, with three families excluded because they frequently have partial length homologs in some lineages).

- Wu, D, Jospin, G, Eisen, JA **(2013)**. Systematic identification of gene families for use as "markers" for phylogenetic and phylogeny-driven ecological studies of bacteria and archaea and their major subgroups. *PLOS ONE*, 8, 10:e77033. doi:[10.1371/journal.pone.0077033](https://dx.doi.org/10.1371/journal.pone.0077033)
- Darling, AE, Jospin, G, Lowe, E, Matsen, FA, Bik, HM, Eisen, JA (2014). PhyloSift: phylogenetic analysis of genomes and metagenomes. PeerJ, 2:e243. doi:[10.7717/peerj.243](https://dx.doi.org/10.7717/peerj.243)


## CheckM

Set of 43 phylogenetically informative marker genes consisting primarily of ribosomal proteins and RNA polymerase domains (and similar to the universal marker set used by PhyloSift).

- Parks, DH, Imelfort, M, Skennerton, CT, Hugenholtz, P, Tyson, GW **(2015)**. CheckM: assessing the quality of microbial genomes recovered from isolates, single cells, and metagenomes. *Genome Res.*, 25, 7:1043-55. doi:[10.1101/gr.186072.114](https://dx.doi.org/10.1101/gr.186072.114)


## GTDB (bac120 and arc122)

Release 86 of the [Genome Taxonomy Database](http://gtdb.ecogenomic.org/).

- Parks, DH, Chuvochina, M, Waite, DW, Rinke, C, Skarshewski, A, Chaumeil, PA, Hugenholtz, P **(2018)**. A standardized bacterial taxonomy based on genome phylogeny substantially revises the tree of life. *Nat. Biotechnol.*, 36, 10:996-1004. doi:[10.1038/nbt.4229](https://dx.doi.org/10.1038/nbt.4229)
- Chaumeil, PA, Hugenholtz, P, Parks, DH **(2018)**. GTDB-Tk: A toolkit to classify genomes with the Genome Taxonomy Database. [https://github.com/Ecogenomics/GTDBTk](https://github.com/Ecogenomics/GTDBTk)
